arch/Kconfig

```diff
@@ -812,6 +817,24 @@ config FPU_SHARING
 
 endmenu
 
+menu "Digital Signal Processing Options"
+
+config DSP
+	bool "This option enables digital signal processing (DSP)"
+	depends on CPU_HAS_DSP
+	help
+	  This option enables DSP and DSP instructions.
+
+config DSP_SHARING
+	bool "DSP register sharing"
+	depends on DSP && MULTITHREADING
+	help
+	  This option enables preservation of the hardware DSP registers
+	  across context switches to allow multiple threads to perform concurrent
+	  DSP operations.
+
+endmenu
```

Comment:
Why do we need `DSP` and `DSP_SHARING`?

So far, in every arch we support, the "DSP"/SIMD registers are considered an extension to the FPU context and we have been doing ok with that model.

Why does ARC need to be special? Why can't it just use `FPU` and `FPU_SHARING` with some ARC-specific Kconfigs for configuring what FPU/SIMD/DSP registers ought to be saved?

Also, having both `FPU` and `DSP` can be very confusing on many existing architectures; for example, on ARM, is NEON considered "FPU" or "DSP" (NEON registers extend the FPU registers on ARM and "NEON context" is considered part of "FPU context").

Reply:
The DSP extensions introduced here are about (efficient HW support for) [fixed point math](https://en.wikipedia.org/wiki/Fixed-point_arithmetic), not floating point. 'hiding' them under FPU doesn't sound right to me.

Reply:
are the "DSP" registers completely separate from the "FPU registers"?

ARM has a "DSP extension" which can be used for both fixed point math and floating point math, but it uses both arch and FPU registers.

If this ARC "DSP" feature has its own completely independent register file not used by the FPU and its instructions at all, and it only supports fixed point math, then that would be something we have not seen before (in fact, ARC might be the only architecture that does that); but, in that case, these `DSP` and `DSP_SHARING` configs should be under ARC.

Reply:
ARC is highly configurable, yes it is possible to have the (fixed point) DSP extensions and no floating point. These are independent features. Let us have a closer look at how it's handled in other architectures and see if it makes sense to further align or make this an ARC only feature.

Reply:
The main question is: "does this DSP feature/unit have its own dedicated register file?"

For ARM, the "DSP extension" uses the arch GP register file for fixed-point math (I mean they are really just integer operations) and VFP register file for floating-point math.

That is, in fact, a quite natural decision to make while you are designing an ISA. So is ARC different in that aspect?

Reply:
ARC has many more DSP features. There are extra MAC registers for the fixed point, as well as some control registers that determine rounding behavior (saturate/round/trunc). Furthermore, there is a nice feature called AGU: address generation unit. That adds extra registers that are automatically load/stored after accessing (with configurable address increments). That eliminates the need for separate load/store instructions. The inner loop of DSP functions becomes 'mac; mac; ... mac' instead of (load; load; mac; store;)*
That's why technically it could be put under FPU, but conceptually that doesn't make sense.

Reply:
Ok, in that case, I really think this needs to be part of `arch/arc` and not a kernel-wide feature. In general, both `DSP` and `DSP_SHARING` are quite vague and confusing when it comes to other architectures and what it is supposed to do and/or achieve is not very clear.

Reply:
The DSP and DSP_SHARING could be introduced locally in arch/arc, how about the status bits in include/zephyr/kernel.h for the DSP sharing? Defining these ARC-local would split the definition, I'd like to avoid that.

Reply:
> how about the status bits in include/zephyr/kernel.h for the DSP sharing? Defining these ARC-local would split the definition, I'd like to avoid that.

Sure, they will need to live in `kernel.h` inside ifdef ARC. wW have `K_SSE_REGS` for x86 there after all.